File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int lft = 0;
        int[] count = new int[128];

        int len = 0 ;

        for(int rght = 0 ; rght < s.length() ; rght++){
            char curr = s.charAt(rght);
            count[curr]++;

            while(count[curr] > 1){
                count[s.charAt(lft)]--;
                lft++;
            }

            len = Math.max(rght - lft + 1 , len);
        }

        return len;
    }
}

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        freq = [0] * 128

        lft = 0
        mx = 0

        for rght in range(len(s)):
            ch = s[rght]
            freq[ord(ch)] += 1

            while freq[ord(ch)] > 1:
                freq[ord(s[lft])] -= 1
                lft += 1

            mx = max(mx , rght - lft + 1)

        return mx